By the way, a la https://github.com/sympy/sympy/issues/20914, I tried to check this using hypothesis. Here is the test I came up with

```py
>>> from hypothesis import given, assume
>>> from hypothesis.strategies import one_of, integers, just
>>> @given(one_of(integers(), just(-oo), just(oo)), one_of(integers(), just(-oo), just(oo)), one_of(integers(min_value=1), integers(max_value=-1)), integers())
... def test_range_as_relational(start, stop, step, val):
...     if start == -oo and stop == oo and step != 1:
...         assume(False)
...     x = Symbol('x')
...
...     r = Range(start, stop, step)
...     rel = r.as_relational(x)
...     assert (val in r) == rel.subs(x, val)
>>> test_range_as_relational()
Falsifying example: test_range_as_relational(
    start=-oo, stop=-oo, step=1, val=0,
)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in test_range_as_relational
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/hypothesis/core.py", line 1142, in wrapped_test
    raise the_error_hypothesis_found
  File "<stdin>", line 8, in test_range_as_relational
  File "./sympy/sets/fancysets.py", line 921, in as_relational
    x >= self.inf if self.inf in self else x > self.inf,
  File "./sympy/sets/sets.py", line 266, in inf
    return self._inf
  File "./sympy/sets/fancysets.py", line 880, in _inf
    raise NotImplementedError
NotImplementedError
```

Once the above issues are fixed, you should try running this function to see if it finds any others.

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -915,7 +915,6 @@ class Range(Set):
             return S.false
         else:
             from sympy.core.mod import Mod
-            from sympy.sets.contains import Contains
             cond = None
             if self.start.is_infinite:
                 if self.stop.is_infinite:
@@ -928,8 +927,7 @@ class Range(Set):
                 a = self.start
             step = abs(self.step)
             cond = Eq(Mod(x, step), a % step) if cond is None else cond
-            return And(Contains(x, S.Integers),
-                       cond,
+            return And(cond,
                        x >= self.inf if self.inf in self else x > self.inf,
                        x <= self.sup if self.sup in self else x < self.sup)
 

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -717,13 +717,12 @@ class Range(Set):
         if not self:
             return S.Zero
         dif = self.stop - self.start
-        if self.has(Symbol):
-            if dif.has(Symbol) or self.step.has(Symbol) or (
-                    not self.start.is_integer and not self.stop.is_integer):
-                raise ValueError('invalid method for symbolic range')
-        if dif.is_infinite:
-            return S.Infinity
-        return Integer(abs(dif//self.step))
+        n = abs(dif // self.step)
+        if not n.is_Integer:
+            if n.is_infinite:
+                return S.Infinity
+            raise ValueError('invalid method for symbolic range')
+        return n
 
     @property
     def is_finite_set(self):
@@ -912,15 +911,27 @@ class Range(Set):
         """Rewrite a Range in terms of equalities and logic operators. """
         if self.size == 1:
             return Eq(x, self[0])
+        elif self.size == 0:
+            return S.false
         else:
             from sympy.core.mod import Mod
-            from sympy.functions.elementary.integers import floor
-            md = self.start % self.step
-            return And(
-                Eq(x, floor(x)),
-                x >= self.inf if self.inf in self else x > self.inf,
-                x <= self.sup if self.sup in self else x < self.sup,
-                Eq(Mod(x, self.step), md))
+            from sympy.sets.contains import Contains
+            cond = None
+            if self.start.is_infinite:
+                if self.stop.is_infinite:
+                    cond = S.true
+                else:
+                    a = self.reversed.start
+            elif self.start == self.stop:
+                cond = S.false  # null range
+            else:
+                a = self.start
+            step = abs(self.step)
+            cond = Eq(Mod(x, step), a % step) if cond is None else cond
+            return And(Contains(x, S.Integers),
+                       cond,
+                       x >= self.inf if self.inf in self else x > self.inf,
+                       x <= self.sup if self.sup in self else x < self.sup)
 
 converter[range] = lambda r: Range(r.start, r.stop, r.step)
 

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -5,7 +5,7 @@ from sympy.core.containers import Tuple
 from sympy.core.expr import Expr
 from sympy.core.function import Lambda
 from sympy.core.logic import fuzzy_not, fuzzy_or, fuzzy_and
-from sympy.core.numbers import oo, Integer
+from sympy.core.numbers import oo
 from sympy.core.relational import Eq
 from sympy.core.singleton import Singleton, S
 from sympy.core.symbol import Dummy, symbols, Symbol

--- a/sympy/sets/tests/test_fancysets.py
+++ b/sympy/sets/tests/test_fancysets.py
@@ -174,6 +174,8 @@ def test_inf_Range_len():
     assert Range(0, -oo, -2).size is S.Infinity
     assert Range(oo, 0, -2).size is S.Infinity
     assert Range(-oo, 0, 2).size is S.Infinity
+    i = Symbol('i', integer=True)
+    assert Range(0, 4 * i, i).size == 4
 
 
 def test_Range_set():
@@ -327,8 +329,8 @@ def test_Range_set():
     assert S(builtin_range(1000000000000)) == Range(1000000000000)
 
     # test Range.as_relational
-    assert Range(1, 4).as_relational(x) == (x >= 1) & (x <= 3) & Eq(x, floor(x)) & Eq(Mod(x,1),0)
-    assert Range(oo, 1, -2).as_relational(x) == (x >= 3) & (x < oo) & Eq(x, floor(x)) & Eq(Mod(x,-2),1)
+    assert Range(1, 4).as_relational(x) == (x >= 1) & (x <= 3) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
+    assert Range(oo, 1, -2).as_relational(x) == (x >= 3) & (x < oo) & Contains(x, S.Integers) & Eq(Mod(x, 2), 1)
 
 
 def test_Range_symbolic():
@@ -383,8 +385,7 @@ def test_Range_symbolic():
     raises(ValueError, lambda: Range(i).inf)
     # as_relational
     raises(ValueError, lambda: sr.as_relational(x))
-    assert ir.as_relational(x) == (
-        x >= i) & Eq(x, floor(x)) & (x <= i + 18) & Eq(Mod(x,2),i%2)
+    assert ir.as_relational(x) == Contains(x,S.Integers) & (x >= i) & (x <= i + 18) & Eq(Mod(x, 2), Mod(i, 2))
     assert Range(i, i + 1).as_relational(x) == Eq(x, i)
     # contains() for symbolic values (issue #18146)
     e = Symbol('e', integer=True, even=True)

--- a/sympy/sets/tests/test_fancysets.py
+++ b/sympy/sets/tests/test_fancysets.py
@@ -329,8 +329,8 @@ def test_Range_set():
     assert S(builtin_range(1000000000000)) == Range(1000000000000)
 
     # test Range.as_relational
-    assert Range(1, 4).as_relational(x) == (x >= 1) & (x <= 3) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
-    assert Range(oo, 1, -2).as_relational(x) == (x >= 3) & (x < oo) & Contains(x, S.Integers) & Eq(Mod(x, 2), 1)
+    assert Range(1, 4).as_relational(x) == (x >= 1) & (x <= 3)  & Eq(Mod(x, 1), 0)
+    assert Range(oo, 1, -2).as_relational(x) == (x >= 3) & (x < oo)  & Eq(Mod(x, 2), 1)
 
 
 def test_Range_symbolic():
@@ -385,7 +385,7 @@ def test_Range_symbolic():
     raises(ValueError, lambda: Range(i).inf)
     # as_relational
     raises(ValueError, lambda: sr.as_relational(x))
-    assert ir.as_relational(x) == Contains(x,S.Integers) & (x >= i) & (x <= i + 18) & Eq(Mod(x, 2), Mod(i, 2))
+    assert ir.as_relational(x) == (x >= i) & (x <= i + 18) & Eq(Mod(x, 2), Mod(i, 2))
     assert Range(i, i + 1).as_relational(x) == Eq(x, i)
     # contains() for symbolic values (issue #18146)
     e = Symbol('e', integer=True, even=True)

--- a/sympy/solvers/tests/test_inequalities.py
+++ b/sympy/solvers/tests/test_inequalities.py
@@ -14,7 +14,7 @@ from sympy.polys.rootoftools import rootof
 from sympy.solvers.solvers import solve
 from sympy.solvers.solveset import solveset
 from sympy.abc import x, y
-from sympy.sets.contains import Contains
+
 from sympy.core.mod import Mod
 
 from sympy.testing.pytest import raises, XFAIL
@@ -418,14 +418,14 @@ def test_integer_domain_relational_isolve():
 
     x = Symbol('x')
     assert isolve(x + 2 < 0, x, domain=S.Integers) == \
-           (x <= -3) & (x > -oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
+           (x <= -3) & (x > -oo) & Eq(Mod(x, 1), 0)
     assert isolve(2 * x + 3 > 0, x, domain=S.Integers) == \
-           (x >= -1) & (x < oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
+           (x >= -1) & (x < oo)  & Eq(Mod(x, 1), 0)
     assert isolve((x ** 2 + 3 * x - 2) < 0, x, domain=S.Integers) == \
-           (x >= -3) & (x <= 0) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
+           (x >= -3) & (x <= 0)  & Eq(Mod(x, 1), 0)
     assert isolve((x ** 2 + 3 * x - 2) > 0, x, domain=S.Integers) == \
-           ((x >= 1) & (x < oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)) | (
-               (x <= -4) & (x > -oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0))
+           ((x >= 1) & (x < oo)  & Eq(Mod(x, 1), 0)) | (
+               (x <= -4) & (x > -oo)  & Eq(Mod(x, 1), 0))
 
 
 def test_issue_10671_12466():

--- a/sympy/solvers/tests/test_inequalities.py
+++ b/sympy/solvers/tests/test_inequalities.py
@@ -14,7 +14,7 @@ from sympy.polys.rootoftools import rootof
 from sympy.solvers.solvers import solve
 from sympy.solvers.solveset import solveset
 from sympy.abc import x, y
-from sympy.functions.elementary.integers import floor
+from sympy.sets.contains import Contains
 from sympy.core.mod import Mod
 
 from sympy.testing.pytest import raises, XFAIL
@@ -418,14 +418,14 @@ def test_integer_domain_relational_isolve():
 
     x = Symbol('x')
     assert isolve(x + 2 < 0, x, domain=S.Integers) == \
-           (x <= -3) & (x > -oo) & Eq(x, floor(x)) & Eq(Mod(x, -1), oo)
+           (x <= -3) & (x > -oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
     assert isolve(2 * x + 3 > 0, x, domain=S.Integers) == \
-           (x >= -1) & (x < oo) & Eq(x, floor(x)) & Eq(Mod(x, 1), 0)
+           (x >= -1) & (x < oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
     assert isolve((x ** 2 + 3 * x - 2) < 0, x, domain=S.Integers) == \
-           (x >= -3) & (x <= 0) & Eq(x, floor(x)) & Eq(Mod(x, 1), 0)
+           (x >= -3) & (x <= 0) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)
     assert isolve((x ** 2 + 3 * x - 2) > 0, x, domain=S.Integers) == \
-           ((x >= 1) & (x < oo) & Eq(x, floor(x)) & Eq(Mod(x, 1), 0)) | (
-                   (x <= -4) & (x > -oo) & Eq(x, floor(x)) & Eq(Mod(x, -1), oo))
+           ((x >= 1) & (x < oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0)) | (
+               (x <= -4) & (x > -oo) & Contains(x, S.Integers) & Eq(Mod(x, 1), 0))
 
 
 def test_issue_10671_12466():

